### ISSUE TITLE
optimize reading / writing GeoPackage

### ISSUE THREAD

**Issue by jessestewart1** [goatsweater/nrn-rrn#19]
**Is your feature request related to a problem? Please describe.**
A large portion of the NRN processing time is spent reading and, especially, writing to disk. The current method of dataframe.iterfeatures() was chosen due to being recommended as a faster alternative as per https://github.com/Toblerity/Fiona/issues/476.

**Describe the solution you'd like**
- Much faster writing to gpkg.
- Somewhat faster reading from gpkg (less of an issue).

**Comment by jessestewart1** [comment]
Managed to find a quicker method for writing to gpkg by using osgeo.ogr.datasource.copylayer. Implementing now.

**Comment by goatsweater** [comment]
Formats that are more focused on speed of reading/writing intermediate data are starting to become easier to implement for geospatial data. For example, there is ongoing work in GeoPandas to implement support for reading/writing to `parquet` files.

Sticking with GPKG makes sense, I think, at least until these formats are better supported by the underlying packages used by the NRN, but once they are available it should make reading and writing layers considerably quicker than the current implementations.

**Comment by goatsweater** [comment]
As part of #10 I've produced an implementation of writing outputs that uses `ogr` directly. The writing of the records seems to be pretty fast, although I did not explicitly time it. While I don't think this can be taken as a true test of write speed, here is what I am seeing re-writing existing GPKGs to new files (ignore the names, this is PE data):

```
Writing NRN_NB_10_0_BLKPASSAGE: 100%|████████████████████████| 148/148 [00:00<00:00, 1166.53it/s]
Writing NRN_NB_10_0_JUNCTION: 100%|██████████████████████████| 10662/10662 [00:08<00:00, 1216.65it/s]
Writing NRN_NB_10_0_TOLLPOINT: 100%|█████████████████████████| 1/1 [00:00<00:00, 77.32it/s]
Writing NRN_NB_10_0_ADDRANGE: 100%|██████████████████████████| 18046/18046 [00:14<00:00, 1283.79it/s]
Writing NRN_NB_10_0_STRPLANAME: 100%|████████████████████████| 6467/6467 [00:04<00:00, 1421.48it/s]
```

One limitation to my approach is that it relies on a call to `.iterfeatures()` to pull the data out of DataFrame and write it to a file. I chose this method of going over the data for a few reasons, but the main one is that a bunch of the formats we support do not allow for bulk writing of data - they write to files one record at a time. While this is generally the slowest way to go over the data, the calls to `ogr` are very fast.

GPKG does have a trick though, in that it can receive the data in bulk and write it in chunks. Going that route would mean a separate code path in my implementation. Ultimately, due to a write mode bug in GeoPandas v0.7 I decided to use a single code path and not treat GPKGs any differently. It may be worth exploring the alternative whenever GeoPandas is updated though, as the bug is already fixed in master.

**Comment by jessestewart1** [comment]
What's the time for roadseg? That's the one dataset which really makes the most difference between difference export methods. Also, do those runtimes include loading the data into ogr? Or is it just the writing to file part?

For spatial datasets I also used ogr directly, but I found that loading the data into an ogr.datasource was just as expensive as writing to file, and, unsurprisingly, it's mainly due to unpacking the geometry column from a shapely object to json. So, I optimized the manual construction of a geojson and passed that to ogr.

For tabular datasets, I found the current implementation of using pandas.to_sql the most efficient.

**Comment by goatsweater** [comment]
Unfortunately my environment causes _reading_ LineString data to produce a segfault (goal version conflicts), so I don't have timings for writing roadseg out. Those timings are only for writes as well. The data is loaded long before I ever started monitoring progress.

Using `.to_sql` would seem the most efficient to me. It leverages the ability to update in bulk if the SQLite session is configured to avoid commits on a per transaction level. My implementation is convenient in that it writes all data the same way (this is also why I went with `ogr` instead of `fiona` - treating attributes tables the same as geometry layers. It is definitely not the most optimized because it effectively leaves all possible speedups to the underlying driver and I have no driver specific options being passed along.

**Comment by jessestewart1** [comment]
Geoparquet support added the other day for future use once it becomes a stable feature of geopandas. Monitoring geopandas PR for reference: https://github.com/geopandas/geopandas/pull/1180. This includes support for pyproj.crs implementation in geopandas >= 0.7.0.

**Comment by martinfleis** [comment]
Hi, sorry for intervening here, but I just noticed

> due to a write mode bug in GeoPandas v0.7

Can you elaborate more? Is that reported/fixed in GeoPandas?

**Comment by goatsweater** [comment]
Hi,

The bug in GeoPandas is that the write mode in `to_file()` had no way to specify a mode other than "w" when writing. This was throwing errors from Fiona that the mode was specified twice, since I was trying to use append.

It has already been fixed as part of geopandas/geopandas#1229.

**Comment by brendan-ward** [comment]
Just adding a few ideas here, based on the link from geopandas.

`parquet` will work great for intermediate files - less so where you need to interoperate with other GIS systems.  The concept is very similar to [geofeather](https://github.com/brendan-ward/geofeather) which I've used for several projects that involve multi-GB data sources.  (geofeather should be superseded by the `parquet` work once that lands in geopandas).  Overall the Arrow based formats (parquet / feather) have worked really well for me.

We are also starting to test out faster ways of working with OGR data sources in a way that is a bit more native to numpy / pandas: [pyogrio](https://github.com/brendan-ward/pyogrio).  It is still very early days there (still working on write functionality, API is not stable, no docs) but I've been using read functionality in multiple projects now with significant speedups over current Fiona / Geopandas loading.  Once that stabilizes a bit more, it would be interesting to compare that to the use of `ogr` as mentioned above.

**Comment by goatsweater** [comment]
I completely agree - using `parquet` for intermediate data will be the fastest option in the future, meaning the only interaction with traditional geospatial formats will be on final output conversion.

Thanks for the heads up on `pyogrio`. I hadn't seen that before, and a bunch of our internals are currently handling the conversion from `geopandas` to `ogr` through `iterfeatures()`. Based on some of your benchmarks we would see a significant speed boost in producing any final outputs once that work gets further along. I think it would be interesting to have a branch in our packager using `pyogrio` at some point to see what kinds of speedups can be realized.
